libnl upstream has introduced some API/ABI changes in version 3.1.  net-snmp
was written against an older version and this patch bridges the changes not
possible in the compatibility header.
---
 net-snmp-5.6.1/agent/mibgroup/mibII/tcpTable.c |   17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

--- a/agent/mibgroup/mibII/tcpTable.c
+++ b/agent/mibgroup/mibII/tcpTable.c
@@ -555,6 +555,7 @@ tcpTable_load(netsnmp_cache *cache, void
 static int
 tcpTable_load_netlink(void)
 {
+	int err;
 	/*  TODO: perhaps use permanent nl handle? */
 	struct nl_handle *nl = nl_handle_alloc();
 
@@ -564,9 +565,9 @@ tcpTable_load_netlink(void)
 		return -1;
 	}
 
-	if (nl_connect(nl, NETLINK_INET_DIAG) < 0) {
-		DEBUGMSGTL(("mibII/tcpTable", "Failed to connect to netlink: %s\n", nl_geterror()));
-		snmp_log(LOG_ERR, "snmpd: Couldn't connect to netlink: %s\n", nl_geterror());
+	if ((err = nl_connect(nl, NETLINK_INET_DIAG)) < 0) {
+		DEBUGMSGTL(("mibII/tcpTable", "Failed to connect to netlink: %s\n", nl_geterror(err)));
+		snmp_log(LOG_ERR, "snmpd: Couldn't connect to netlink: %s\n", nl_geterror(err));
 		nl_handle_destroy(nl);
 		return -1;
 	}
@@ -579,9 +580,9 @@ tcpTable_load_netlink(void)
 	struct nl_msg *nm = nlmsg_alloc_simple(TCPDIAG_GETSOCK, NLM_F_ROOT|NLM_F_MATCH|NLM_F_REQUEST);
 	nlmsg_append(nm, &req, sizeof(struct inet_diag_req), 0);
 
-	if (nl_send_auto_complete(nl, nm) < 0) {
-		DEBUGMSGTL(("mibII/tcpTable", "nl_send_autocomplete(): %s\n", nl_geterror()));
-		snmp_log(LOG_ERR, "snmpd: nl_send_autocomplete(): %s\n", nl_geterror());
+	if ((err = nl_send_auto_complete(nl, nm)) < 0) {
+		DEBUGMSGTL(("mibII/tcpTable", "nl_send_autocomplete(): %s\n", nl_geterror(err)));
+		snmp_log(LOG_ERR, "snmpd: nl_send_autocomplete(): %s\n", nl_geterror(err));
 		nl_handle_destroy(nl);
 		return -1;
 	}
@@ -593,8 +594,8 @@ tcpTable_load_netlink(void)
 
 	while (running) {
 		if ((len = nl_recv(nl, &peer, &buf, NULL)) <= 0) {
-			DEBUGMSGTL(("mibII/tcpTable", "nl_recv(): %s\n", nl_geterror()));
-			snmp_log(LOG_ERR, "snmpd: nl_recv(): %s\n", nl_geterror());
+			DEBUGMSGTL(("mibII/tcpTable", "nl_recv(): %s\n", nl_geterror(len)));
+			snmp_log(LOG_ERR, "snmpd: nl_recv(): %s\n", nl_geterror(len));
 			nl_handle_destroy(nl);
 			return -1;
 		}
